### PR TITLE
Security report crash bugfix

### LINKED ISSUE #2600: Security report crashes if no agent is done running
## Describe the bug
If you try to view the security report before any agent is done running, it crashes. You have to wait for one agent to finish running before viewing it.

## To Reproduce
1. Run the Island and Agent.
2. Try viewing the security report before the Agent finishes its run.
3. See it crash.

## Expected behavior
It shouldn't crash.

## Tasks
- [ ] Fix it (0.25d) - @shreyamalviya

### PR DESCRIPTION
**Pull request by shreyamalviya**
# What does this PR do?

Fixes #2600 

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [x] Is the TravisCI build passing?
* [ ] ~Was the CHANGELOG.md updated to reflect the changes?~
* [ ] ~Was the documentation framework updated to reflect the changes?~
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [ ] ~Added relevant unit tests?~
* [x] Have you successfully tested your changes locally? Elaborate:
    > Tested by running the Island
* [x] If applicable, add screenshots or log transcripts of the feature working

https://user-images.githubusercontent.com/44770317/204231467-6c1d29bc-ae19-4b7c-8188-817d199a3b3a.mp4

### CODE REVIEW

File: monkey/monkey_island/cc/services/reporting/report.py [hunk 1]
```diff
@@ -95,13 +95,20 @@ def get_first_monkey_time(cls):
...
         return max(agents, key=lambda a: a.stop_time).stop_time
 
     @staticmethod
     def get_monkey_duration():
-        delta = ReportService.get_last_monkey_dead_time() - ReportService.get_first_monkey_time()
+        last_monkey_dead_time = ReportService.get_last_monkey_dead_time()
+        if last_monkey_dead_time == 0:
+            return 0
+
+        delta = last_monkey_dead_time - ReportService.get_first_monkey_time()
```
Comment (ilija-lazoroski) [inline monkey/monkey_island/cc/services/reporting/report.py:111]: Probably we should do the same when we get the latest timestamp from events.

File: monkey/monkey_island/cc/ui/src/components/report-components/SecurityReport.js [hunk 2]
```diff
@@ -226,6 +226,18 @@ class ReportPageComponent extends AuthComponent {
     return Object.keys(this.state.report).length === 0;
   }
 
+  getMonkeyDuration() {
+    if (this.state.report.overview.monkey_duration !== 0) {
+      return <>
+        After <span
+        className='badge badge-info'>{this.state.report.overview.monkey_duration}</span>, all monkeys finished
+        propagation attempts.
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/ui/src/components/report-components/SecurityReport.js:234]: This says "all monkeys finished propagation attempts." but they didn't. This is show with the latest agent stop time, even though agents are still running.

File: monkey/monkey_island/cc/ui/src/components/report-components/SecurityReport.js [hunk 3]
```diff
@@ -226,6 +226,18 @@ class ReportPageComponent extends AuthComponent {
     return Object.keys(this.state.report).length === 0;
   }
 
+  getMonkeyDuration() {
+    if (this.state.report.overview.monkey_duration !== 0) {
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/ui/src/components/report-components/SecurityReport.js:230]: We don't need to check the duration, just call `didAllAgentsShutdown`

File: monkey/monkey_island/cc/services/reporting/report.py [hunk 4]
```diff
@@ -95,13 +95,20 @@ def get_first_monkey_time(cls):
...
-        agents = filter(lambda a: a.stop_time is not None, cls._agent_repository.get_agents())
+        agents = list(filter(lambda a: a.stop_time is not None, cls._agent_repository.get_agents()))
+
+        if not agents:
+            return 0
 
         return max(agents, key=lambda a: a.stop_time).stop_time
 
     @staticmethod
     def get_monkey_duration():
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/services/reporting/report.py:106]: We should probably move this and `get_last_monkey_dead_time` to the front end

**Comment by VakarisZ** [comment]
What we should be doing is gradually refactoring the report to the front end. Since that's not our main focus for now, we can do the bare minimum to solve this issue. So there are two sollutions:
- Move the duration calculation to the front end. Should be pretty easy, we already have all of the agents in the state, just order them by timestamp and subtract first timestamp from the last
- Keep the duration calculation on the back end, but fix to return `None` instead of 0. 

Either way we need to only display the duration if all agents have died.

File: monkey/monkey_island/cc/ui/src/components/report-components/SecurityReport.js (hunk 3 above)
Comment (shreyamalviya) [reply monkey/monkey_island/cc/ui/src/components/report-components/SecurityReport.js:230]: The promise doesn't resolve in time and I don't think we should crowd this with state for everything when we're going to rework it anyway. I'm leaving it as it is.

Comment (VakarisZ) [reply monkey/monkey_island/cc/ui/src/components/report-components/SecurityReport.js]: I'd anticipate that report will need to show the duration even after the rework. Yes, that method queries the island even though we already have the list of agent. We should separate the request part from parsing the agent list part.

File: monkey/monkey_island/cc/services/reporting/report.py [hunk 5]
```diff
@@ -94,14 +95,28 @@ def get_first_monkey_time(cls):
...
-        agents = filter(lambda a: a.stop_time is not None, cls._agent_repository.get_agents())
+    def get_last_monkey_dead_time(cls) -> Optional[datetime]:
+        agents = cls._agent_repository.get_agents()
+        if not agents:
+            return None
+
+        # TODO: Make sure that the case where an agent doesn't have a stop time
+        #       because it doesn't send a shutdown event is solved after #2518.
+        #       Till then, if an agent doesn't have stop time, the total run duration
+        #       won't be shown in the report.
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/services/reporting/report.py:106]: So now we'll have to go back here and fix the code that doesn't belong here in the first place?

**Review by VakarisZ — APPROVED** [review]

**Review by mssalvatore — APPROVED** [review]

File: monkey/monkey_island/cc/services/reporting/report.py (hunk 5 above)
Comment (shreyamalviya) [reply monkey/monkey_island/cc/services/reporting/report.py:106]: We won't necessarily have to fix it, and this was an issue (it returned the monkey run duration even if agents were still running) that existed before this PR.

**Pull request merged**